Allow displaying custom columns in stylesheet

pbiron opened the pull request:
### Issue Number

#152

### Description

This is a simplified version of the solution in #153.

Introduces the `core_sitemaps_stylesheet_columns` filter that allows plugins to change the columns rendered by the stylesheet.  This new filter is analogous to `manage_posts_columns` (and all the other `WP_List_Table` "columns" filters).

Unlike #153, the stylesheet generated by this PR does **not** try to render columns for **any** child of `Q{http://www.sitemaps.org/schemas/sitemap/0.9}url`, just those returned by `core_sitemaps_stylesheet_columns`, which means the stylesheet is greatly simplified from the one in that PR...and only _slightly_ more complex than the current stylesheet.

Like #153, it also introduces the `esc_xml()` and `esc_xml__()` functions.

There is a pretty extensive unit test for the rendered HTML.  The `Test_Core_Sitemaps_Renderer::loadXML()` method (introduced in #157) is also modified so that it can load XML from a file, which is needed by `test_stylesheet_html()` (the details of why that is necessary are explained in an inline comment in the test).

The index stylesheet is also cleaned up a bit, although custom columns are **not** allowed for it.  Why?  Because there are currently no hooks that allow plugins to add extension elements to the sitemap index.  If/when those hooks are added, the index stylesheet can easily be adapted to allow custom columns.

### Screenshots (before and after if applicable)

#### No custom columns

![no-custom-columns](https://user-images.githubusercontent.com/3824560/80292961-310b2400-8718-11ea-8cc3-9107ff01194c.png)

#### Change the text used for the URL column

```
add_filter(
	'core_sitemaps_stylesheet_columns',
	function( $columns ) {
		$columns['http://www.sitemaps.org/schemas/sitemap/0.9']['loc'] = 'Permalink';

		return $columns;
	}
);
```
![rename-url](https://user-images.githubusercontent.com/3824560/80292996-b1318980-8718-11ea-9e94-ac51d73b0e69.png)

#### Add Change Frequency

```
add_filter(
	'core_sitemaps_posts_url_list',
	function ( $url_list ) {
		foreach ( $url_list as &$url ) {
			$url['changefreq'] = 'weekly';
		}
		
		return $url_list;
	}
);
add_filter(
	'core_sitemaps_stylesheet_columns',
	function( $columns ) {
		$columns['http://www.sitemaps.org/schemas/sitemap/0.9']['changefreq'] = 'Change Frequency';

		return $columns;
	}
);
```
![changefreq](https://user-images.githubusercontent.com/3824560/80293046-3321b280-8719-11ea-86de-ea6b372365d3.png)

#### Reorder columns

```
// I don't know why a plugin would want to do this;
// I also don't know why they'd want to reorder columns
// on WP_Posts_List_Table, but manage_posts_columns allows them to do that :-)
add_filter(
	'core_sitemaps_posts_url_list',
	function ( $url_list ) {
		foreach ( $url_list as &$url ) {
			$url['changefreq'] = 'weekly';
		}
		
		return $url_list;
	}
);
add_filter(
	'core_sitemaps_stylesheet_columns',
	function( $columns ) {
		$columns['http://www.sitemaps.org/schemas/sitemap/0.9'] =
			array(
				'changefreq' => 'Change Frequency'
			) +
			$columns['http://www.sitemaps.org/schemas/sitemap/0.9'];

		return $columns;
	}
);
```
![reorder-columns](https://user-images.githubusercontent.com/3824560/80293150-07eb9300-871a-11ea-97a6-09ad93e94297.png)

#### Add custom to sitemap element & column to stylesheet output

```
add_filter(
	'core_sitemaps_posts_url_list',
	function ( $url_list ) {
		foreach ( $url_list as &$url ) {
			// changefreq won't appear in the rendered HTML
			// because isn't not added by core_sitemaps_stylesheet_columns below.
			$url['changefreq'] = 'weekly';
			$url['foo']        = 'bar';
		}
		
		return $url_list;
	}
);
add_filter(
	'core_sitemaps_stylesheet_columns',
	function( $columns ) {
		$columns['http://www.sitemaps.org/schemas/sitemap/0.9']['foo'] = 'Custom Column';

		return $columns;
	}
);
```
![custom-column](https://user-images.githubusercontent.com/3824560/80293200-8f390680-871a-11ea-9a15-e7265f3ab9c1.png)

#### Style columns.
```
// again, not sure why a plugin would want to do this, but
// they can style columns (and body cells) in WP_List_Tables.
add_filter(
	'core_sitemaps_posts_url_list',
	function ( $url_list ) {
		foreach ( $url_list as &$url ) {
			$url['changefreq'] = 'weekly';
		}
		
		return $url_list;
	}
);
add_filter(
	'core_sitemaps_stylesheet_columns',
	function( $columns ) {
		$columns['http://www.sitemaps.org/schemas/sitemap/0.9']['changefreq'] = 'Change Frequency';

		return $columns;
	}
);
add_filter(
	'core_sitemaps_stylesheet_css',
	function( $css ) {
		// garish colors, to make sure the styling is obvious.
		$css .= <<<EOF
#sitemap__table thead th.loc[class~="http://www.sitemaps.org/schemas/sitemap/0.9"] { color: red; }
#sitemap__table tbody td.loc[class~="http://www.sitemaps.org/schemas/sitemap/0.9"] a { color: green; }
#sitemap__table thead th.changefreq[class~="http://www.sitemaps.org/schemas/sitemap/0.9"] { color: orange; }
#sitemap__table tbody td.changefreq[class~="http://www.sitemaps.org/schemas/sitemap/0.9"] { color: blue; }

EOF;

		return $css;
	}
);
```
![style-columns](https://user-images.githubusercontent.com/3824560/80293250-19816a80-871b-11ea-80e8-e1c12f0d8543.png)

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Enhancement (change which improves an existing feature. E.g., performance improvement, docs update, etc.)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test

I've tested this in Chrome, Firefox, Edge and IE (11) on Windows. Would really appreciate folks testing it on Mac & Linux with as many different browsers as you can!!

See the example code above each screenshot for how to test.

### Acceptance criteria
- [X] My code follows WordPress coding standards.
- [X] I have performed a self-review of my own code.
- [X] If the changes are visual, I have cross browser / device tested.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] My changes generate no new warnings.
- [X] I have added test instructions that prove my fix is effective or that my feature works.

pbiron committed:
Introduce the `core_sitemaps_stylesheet_columns` filter to allow plugins to change the columns rendered by the stylesheet (and correspoding changes to the generation of the XSLT of the stylesheet).

pbiron committed:
Add unit test for the HTML rendered by the stylesheet.

Also modifies Test_Core_Sitemaps_Renderer::loadXML() so that it can load XML from a file.

pbiron commented:
This is ready to be reviewed.

I made it a draft for the following reasons:

* As is, the `core_sitemaps_stylesheet_columns` filter is used by the stylesheet for the sitemap for **all** providers
    * There really should be different filters for each provider (e.g., `core_sitemaps_stylesheet_posts_post_columns`, `core_sitemaps_stylesheet_taxonomies_category_columns`, etc), just like there are different `manage_XYZ_columns` filters for different list tables
    * However, at the time the stylesheet is generated, there is no way to tell which provider's sitemap it is being used with
    * I have a working solution to that problem
       * Add provider-specific rewrite rules for the stylesheet (e.g., for `wp-sitemap-posts-post.xsl`, `wp-sitemap-users.xsl`)
       * Change `Core_Sitemaps_Renderer::get_sitemap_stylesheet_url()` to take the `sitemap` and `sub_type` query_vars into account when generating the stylesheet URL
    * But I'm wondering if anyone has ideas about other possible solutions to _that_ problem
        * And I wanted the initial reviews of this to focus on reviewing the XSLT :-)
    * Also, given #150, the exact way the additional rewrite rules would be added is likely to change...so I didn't want to complicate things at this point
* The `core_sitemaps_stylesheet_columns` filter (and the way it is used to generated the XSLT) anticipates _some_ solution to #151 (i.e., that extension elements in the sitemap are in extension namespaces, and not in the sitemaps namespace as is currently the case)
    * If a solution to #151 will not be in the merge proposal, then this can be simplified even further

pbiron commented:
@swissspidy the travis failure is caused by the same WPCS failures in #153 and I still don't know how to suppress them.

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+	 * @return string Translated text.
+	 */
+	function esc_xml__( $text, $domain = 'default' ) {
+		return esc_xml( translate( $text, $domain ) );
```
```suggestion
		return esc_xml( translate( $text, $domain ) ); // phpcs:ignore WordPress.WP.I18n
```

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param string $text
```
Parameter needs a description, e.g. "Text to be escaped"

swissspidy replied:
```suggestion
	 * @param string $text Text to escape.
```

pbiron replied:
this and the other PHPCS changes you suggested are now done.

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+	 * @param string $text
+	 * @return string
+	 */
+	function esc_xml( $text ) {
```
```suggestion
	function esc_xml( $text ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+		 * @param string $safe_text The text after it has been escaped.
+		 * @param string $text      The text prior to being escaped.
+		 */
+		return apply_filters( 'esc_xml', $safe_text, $text );
```
```suggestion
		return apply_filters( 'esc_xml', $safe_text, $text ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+	 *                       Default 'default'.
+	 * @return string Translated text.
+	 */
+	function esc_xml__( $text, $domain = 'default' ) {
```
```suggestion
	function esc_xml__( $text, $domain = 'default' ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```

pbiron committed:
Coding standards: add description to $text param in DocBlock of esc_xml() and suppress other PHPCS warnings for esc_xml() and esc_xml__().

swissspidy commented on inc/class-core-sitemaps-stylesheet.php:
```diff
@@ -237,4 +328,40 @@ protected function get_stylesheet_css() {
+	protected function get_stylesheet_columns() {
+		$default_columns = array(
+			'http://www.sitemaps.org/schemas/sitemap/0.9' => array(
+				'loc' => esc_xml__( 'URL', 'core-sitemaps' ),
```
Doesn't `__()` suffice since `esc_xml` is called in the loop anyway?

pbiron committed:
Merge branch 'master' into enhancement/152-stylesheet-custom-columns

swissspidy commented on inc/functions.php:
```diff
@@ -87,3 +87,51 @@ function core_sitemaps_get_max_urls( $type = '' ) {
+	 * @param string $text Text to escape.
+	 * @return string
+	 */
+	function esc_xml( $text ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```
Perhaps we can split this up into its own PR?

pbiron committed:
Merge branch 'master' into enhancement/152-stylesheet-custom-columns

pbiron committed:
Correct unresolved merge conflict.

pbiron committed:
Update filter name to use the new naming conventions.

pbiron committed:
Remove the esc_xml() and esc_xml__() functions.

A separate PR (#192) was opened to add those.  When that gets merged, this will be updated to use them.

pbiron committed:
Allow sitemap-specific custom columns to be rendered.

pbiron committed:
Be consistent in the order of @type and @href in the xml-stylesheet PIs for sitemaps and index.  Also fixes WPCS.

pbiron committed:
Update unit tests for changes in stylesheet URLs.

pbiron committed:
Cast $page to string in WP_Sitemaps_Provider::get_sitemap_url().

pbiron commented:
This is now ready for review, and it now allows different custom columns to be rendered for different sitemaps.  There is no change to the XSLT from what was already in this PR.  

Here's a list of the recent changes:

1. adds rewrite rules for stylesheets that are analogous to those that already exist for sitemaps themselves
   a. `https://example.com/wp-sitemap-posts-post-1.xml` will have stylesheet URL `https://example.com/wp-sitemap-posts-post.xsl`, while `https://example.com/wp-sitemap-users-1.xml` will have stylesheet URL `https://example.com/wp-sitemap-users.xsl`, etc
   b. note that there is no "paged" query arg for the stylesheet rewrite rules as there is is for sitemap rules

2. delays generating the xml-stylesheet processing instruction until `WP_Sitemaps_Renderer::get_sitemap_xml()`, so that the sitemap the stylesheet for is known (and so that different stylesheet URLs can be generated)
   a. there is no longer any need for the `$stylesheet` and `$stylesheet_index` properties of `WP_Sitemaps_Renderer`, so those are removed

3. passes sitemap and sitemap sub-type parameters to the `wp_sitemaps_stylesheet_columns` filter so that plugins can add different columns depending on the sitemap being rendered
   a. also passes those to the `wp_sitemaps_stylesheet_css` filter, so that sitemap-specific CSS can also be added

4. extracts the logic in `WP_Sitemaps_Provider::get_sitemap_url()` into a helper function so that it can be called both from there and from `WP_Sitemaps_Renderer::get_sitemap_stylesheet_url()`

5. removes the `esc_xml()` and `esc_xml__()` functions, in favor of #192.  When that gets merged this can be updated to use them

6. removes some insignificant whitespace in the XSLT

pbiron committed:
Merge branch 'master' into enhancement/152-stylesheet-custom-columns

pbiron commented:
gonna close this out.
